### ISSUE TITLE
Document and/or improve tools for safely interacting with external memory

### ISSUE THREAD

**Issue by DemiMarie** [KhronosGroup/Vulkan-Docs#2408]
The Vulkan specification requires that applications not cause data races:

> Applications **must** ensure that no data races occur during the execution of their application.

However, this requirement cannot be satisfied for imported buffers, because the exporting process may modify the memory at any time.  While a correct non-malicious exporter will not do so, a data race caused by a buggy or malicious exporter must not be able to harm the importing application.  Due to memory bandwidth limitations, I suspect it is unreasonable to expect importing programs to perform a defensive copy using one of the copy commands.

**Comment by cubanismo** [comment]
This is beyond the level of security provided by the current external memory extensions. The external memory extensions guarantee only that the applications won't be able to clobber each-others' non-shared resources or cause the other application to terminate IIRC. Beyond that, some level of mutual trust is indeed assumed. If you have an environment that requires stronger protections and can provide them, you could define additional extensions defining the additional guarantees there.

**Comment by DemiMarie** [comment]
From what I understand, Wayland compositors are expected to not need to have any trust in their clients, and implementations of the dmabuf extension are expected to ensure this.  Furthermore, Wayland compositors do not actually perform any copies of imported buffers.  This means that the guarantees provided by this extension are weaker than both what is assumed by its users and what is expected from its implementors.  In particular, data races are (to the best of my understanding) undefined behavior, which means anything can happen.  This means that the current specification is inconsistent.

While it would be possible to add a new extension to codify this, I think it would be more useful to update the current spec to match what is both assumed and (to the best of my knowledge) provided.

I’m CCing @marmarek (from Qubes OS), @emersion (wlroots lead maintainer), @gfxstrand (Linux graphics stack developer), and @robclark (ChromeOS).

**Comment by DemiMarie** [comment]
> This is beyond the level of security provided by the current external memory extensions. The external memory extensions guarantee only that the applications won't be able to clobber each-others' non-shared resources or cause the other application to terminate IIRC.

What I find confusing is that the Vulkan specification seems to contradict itself here.  Specifically:

> Applications **must** ensure that no data races happen during the execution of their application.

To the best of my understanding, the “must” implies that if it is not satisfied (a data race happens), the behavior is undefined and might include application termination or memory corruption.  This seems to contradict this sentence from §11.2.3:

> However, implementations **must** perform sufficient validation of external handles and payloads to ensure that the operation results in a valid memory object which will not cause program termination, device loss, queue stalls, or corruption of other resources when used as allowed according to its allocation parameters.

and that is why I am confused: §11.2.3 implies that the imported buffer object _can_ be safely used, yet the prohibition on data races seems to contradict that.

**Comment by gfxstrand** [comment]
This is indeed a tricky thing to pin down but it's also fine in practice. Take, for instance, a shader that does the the following:
```
layout (set = 0, binding = 1) uniform block S {
   uint32_t x;
} ssbo;

void main() {
    ssbo.x += 1;
}
```
and assume that's a compute shader run once (so no parallelism in our process) and that the SSBO lives in shared memory. In the presence of a data race (whether by this process or another) we can make no guarantees about the final value in that SSBO. It's not an atomic. It does a load, add, and store. So if someone else comes in between the load and the store and messes with the data, we can't guarantee anything anymore.

Or more of an interesting case for compositors... Suppose someone renders at the same time you're texturing. What data are you going to get? We don't know. We can't know. There's nothing we can do to protect against that. We don't even know if things will behave consistently because of cache behavior which normally wouldn't be a problem thanks to pipeline barriers but, again, data races make all those guarantees go away.

But in practice how bad will it be? You'll get the wrong data composited. That's about it. If there's some image compression going on, things might look more funky than just a clean tear line but it's not going to take down the GPU or anything.

That said, there is no general guarantee we can really provide. If you have a shader that loops based on some bit of data it thinks it's incrementing but isn't because of data races, we can't guarantee that shader will terminate. If you're crawling a tree data structure in some bit of shared memory, we can't guarantee that the pointers/offsets you load will be valid and that you won't end up off in the weeds or faulting. But this is no different from any other shared memory scenario. The C spec doesn't provide any guarantees in the presence of data races, either, and we use shared memory IPC all the time. This is no different.

**Comment by DemiMarie** [comment]
> The C spec doesn't provide any guarantees in the presence of data races, either, and we use shared memory IPC all the time. This is no different.

I actually think this is a bit different.  There are well-known techniques in C for avoiding races with minimal performance penalty, and programmers who can't use them (perhaps because they need SIMD) can fall back to assembly language.  In the GPU world, these techniques are (to the best of my knowledge) not well-known at all, and e.g. wlroots’s Vulkan renderer doesn’t seem to use them.

That said, what keeps security people like me up at night is the “halt and catch fire” semantics of data races.  For instance, I expect it to be not uncommon that the results of reading a shared buffer are never used for control flow or address generation and only affect a subset of the output.  A naive compilation would preserve this property even in the presence of data races, since (as you mentioned above) data races do _not_ have “halt and catch fire” semantics at the machine level.  The problem is if an overly-clever compiler generates stuff like:

```c
	uint8_t *buffer;
	if (*buffer != *buffer)
		clobber_memory();
```

This is obviously a toy example, but I think this makes the broader point clear: people whose job is to think about what _could_ happen prefer to rely on what the spec guarantees, or at least on “if you break this you break the Linux kernel.”

I think there are a few guarantees that would be useful, in what I expect are roughly increasing order of controversy:

1. If a process uses atomic loads and stores on a shared buffer while another process uses non-atomic operations, only the process that uses non-atomic operations is considered to have caused a data race.  My understanding is that Rust intends to provide this guarantee for CPU accesses.
2. If a buffer is accessed using one of the copy commands while another process is reading the buffer, the behavior of the copy command is not affected.
3. If a buffer is accessed using one of the copy commands while another process is writing the buffer, the copy command will copy junk data but not cause corruption or device loss.
4. If the result of a non-atomic load is never used for address computations or control flow, computations involving this result may produce junk values, but memory corruption will not result.

**Comment by cubanismo** [comment]
Discussed this a bit in the Vulkan System Integration ("SI") TSG this week. The below is extrapolated from the general consensus there that probably is a real issue to address here, but the SI TSG is probably not the right group to address it:

- There is motivation for providing fixes akin to what @DemiMarie suggests when using external memory. The exact form of the fixes is unclear, but it is reasonable to suggest the addition of this data race language regressed the ability to use external memory in Vulkan as originally intended. When external memory and external semaphore support was added, we attempted to enable the sharing of Vulkan memory across trust boundaries to some extent, in that a lesser-privileged application sharing memory with a more-privileged application should not be able to cause the more-privileged application to terminate or hang. There is some language for semaphore import that attempts to define the level of guarantees provided:
   > When importing a semaphore payload, it is the responsibility of the application to ensure the external handles meet all valid usage requirements. However, implementations must perform sufficient validation of external handles to ensure that the operation results in a valid semaphore which will not cause program termination, device loss, queue stalls, or corruption of other resources when used as allowed according to its import parameters, and excepting those side effects allowed for violations of the valid semaphore state for wait operations rules. If the external handle provided does not meet these requirements, the implementation must fail the semaphore payload import operation with the error code VK_ERROR_INVALID_EXTERNAL_HANDLE.
- Solving this requires changes to core spec language, or at the very least, carving out exceptions from the cited memory model spec language (For reference, it's from "Appendix B: Memory Model", under the heading "Data Race") for external memory-backed resources, but I see no clear reason that external memory motivating the fix necessitates limiting the fix to external memory. It would probably just complicate things.
- The language changes necessary to solve this are beyond the scope of the SI TSG participants. It is likely the memory model TSG would be the best group to craft any necessary spec updates. If needed, the SI TSG could assist with wording to scope the changes to external memory, if so desired.

Hence, I'm removing the system integration label, and adding the memory model TSG label. Let me know if there's a desire to coordinate work on this issue between the two groups. @oddhack, could you route this appropriately?

**Comment by DemiMarie** [comment]
Thanks @cubanismo!

I strongly recommend an approach that blesses what existing compositors already do, rather than requiring changes to existing compositor shaders.  The rationale is that compositors are already complex enough, and fixing the spec seems much better than requiring ecosystem-wide changes.

**Comment by oddhack** [comment]
> Hence, I'm removing the system integration label, and adding the memory model TSG label. Let me know if there's a desire to coordinate work on this issue between the two groups. [@oddhack](https://github.com/oddhack), could you route this appropriately?

I'm not sure who's chairing the MM TSG at the moment, but I expect @r-potter does. Best to assign to whomever that is. We could also mention it on the next maintenance call.

**Comment by Tobski** [comment]
Can someone describe the exact series of events that would lead to the type of race described here? I'm trying to figure out where the supposed race is, and I can't find one. The only possibility I see is if the external entity is for some reason writing data while the Vulkan implementation has ownership, which isn't something we had intended to allow originally.

Also, @gfxstrand, please don't recommend to people that data races aren't going to be bad in practice! Yes, there's a bunch of stuff that technically _could_ be ok, but the spec says what it says because we haven't yet defined what is or is not ok - (which we are working on!), until then, like with C/C++, please just assume if you have a data race, stuff might start blowing up.

**Comment by DemiMarie** [comment]
@Tobski: The possibility you are considering is exactly what could lead to this race.  It’s not a situation that will happen if both the application and the external entity are correct.  However, a malicious external entity may _intentionally_ violate the Vulkan or system integration specs in an attempt to deceive the user or escalate its privileges.

**Comment by Tobski** [comment]
@DemiMarie thanks for the clarification! So it's not so much about whether a data race occurs or not - it's more about limiting the damage if so-called undefined behavior happens? If so that's a much more tractable problem, and doesn't need memory model shenanigans, but security guarantees. We already have something like that for protected memory in the spec, so we could probably repurpose it. I guess the question is what security guarantees you're after?

Something like "can't leak memory cross process" and "can't cause termination in another process" probably?

**Comment by cubanismo** [comment]
To be clear, the Vulkan spec never tried to provide extensive security guarantees with external memory. It was always possible to, as you say, potentially deceive the user by clobber in-flight use of memory by the compositor by using it from a malicious client of the compositor at the same time.

What changed with the MM appendix is that such usage became explicitly undefined, whereas we took some pains to guarantee it wouldn't cause device lost or crashes when originally adding external memory, so the wording added by MM could be considered a regression to some extent. It was ill-defined before, but less defined after.

**Comment by gfxstrand** [comment]
For general external memory access, there's likely not much we can say. In the case I cited above, you can very much run into issues if the maliscious client is trying to trick the compositor. The more complex the compositor's access and the more it does with that data, the more likely it is to run into problems. However, this is fundamentally no different from any other shared memory protocol. The moment you touch a shared memory buffer from C, you're potentially opening yourself up to undefined behavior according to the C/C++ specs. Vulkan is no different there.

But we probably can limit the damage somewhat. For instance, we might be able to say that if a texture operation encounters a data race with something writing to the texture's memory, the worst case is that you get garbage back from the texture unit. This doesn't say "data races are safe" or "data races don't happen", it just limits the damage in the case that compositors actually care about. Undefined values are a lot easier to code defensively against than full undefined behavior.

And I disagree with @cubanismo a bit here. I don't think the MM appendix made it any less defined. We just started being honest about it. The danger was always there.

**Comment by DemiMarie** [comment]
Yes, I need security guarantees for this.  The guarantees I need are fairly strong, though: if a client of a compositor races it cannot affect what the user sees from other clients.  I’m not sure what that translates into in Vulkan terms, but @valpackett probably does.

Something of the form “if there is a data race on external memory, the behavior is as if each use of the object the race occurs results in a non-deterministic value that could have been obtained without a race” should be sufficient.  There are probably weaker guarantees that also work, but I don’t consider myself sufficiently knowledgeable with Vulkan to state what exactly they are.

The application-layer guarantee that I am looking for is that if one client of a compositor races, it only affects the part of the screen derived from its own buffers, not the parts derived from other clients.  Furthermore, this should not require every compositor to rewrite their shaders to only use atomic operations, especially because I believe not every operation has an atomic version.  It also should not require expensive mitigations like performing a defensive copy of client-provided buffers before using them as pixel contents.

My understanding is that Wayland display compositors are by far the most common users of Vulkan external memory.  Since they _do_ need the strong application-layer guarantee I mentioned above, I think it is best for the Vulkan specification to make it easy for them to provide it.

**Comment by Robert-Qualcomm** [comment]
I'm a bit late to this discussion due to some github issues. I chair the memory model TSG and we will certainly discuss what we can do here, but I think the options are limited.

The Vulkan memory model follows the philosophy of the C++11 memory mode. As far as the shader is concerned, data races result in UB. Since the compiler and hardware can disregard any execution path that results in a data race, the behaviour is completely undefined. 

In practice, a data race can result in several possible behaviours including tearing (reading a combination of an old value and new value), invariance violation (the compiler assumes a memory location will not be written and is therefore constant, but a data races breaks this assumption), illegal values e.g. if boolean is mapped to integer values '0' for false and '1' for true but a data race can result in other values being read. Mitigations exist for these and other specific failure modes, but the possible behaviour in the presence of data races is unbounded. In summary, we can't know what some future implementation might do so we can't put in fixes for all possible errant behaviours.

As far as I know, there are only two approaches to constraining UB:

1. Replace all normal loads and stores by atomic loads and stores. This works because the memory model allows the user to designate conflicts as either legal (i.e. atomic) or illegal (data race). The reason to categorize conflicts as data races enables certain optimizations so mapping all loads and stores to atomics incurs a performance cost.

2. Constrain the undefined behaviour by adding functionality to the Vulkan spec as a whole. For example, if the problematic behaviours are OOB accesses and deadlocks, the shader can be sandboxed and a watchdog timer added. The robustness spec does something similar for OOB accesses.

**Comment by mahkoh** [comment]
The memory in this context is accessed by importing it as a VkImage. Access involves layout transitions of the VkImage which I assume counts as memory access too. The image is used as a color attachment or for sampling or with vkCmdBlitImage or with vkCmdCopyImage and so on. All of the operations that are not inside a shader would need to be made safe too.

**Comment by DemiMarie** [comment]
What about declaring data races on external memory to not be undefined?  Implementations could either assume that data races could happen everywhere, or only assume that data races will not happen if external memory is not involved.

**Comment by gfxstrand** [comment]
That doesn't work. Certain aspects of undefined behavior are because compiler optimizations make assumptions about the consistency of memory operations. We have no idea what resources come from external memory and what don't at compile time so there's no way to can defend against those cases based on whether or not something is external. It'll have to be a general relaxation or some guarantee we can make in the presence of extra shader flags like `coherent` or `volatile`, not based on whether or not something is external.

**Comment by DemiMarie** [comment]
What about a flag set at device creation time, or perhaps command buffer creation time?  It’s a big hammer, but it should do the job.

I considered suggesting that the implementation recompile if needed, but my understandint is that Vulkan Vulkan tries to avoid implicit recompilation.

**Comment by gfxstrand** [comment]
Yeah, on-the-fly re-compilation is right out. Device flags aren't great, either. Something at compile time might work but it's not clear what it would do and what the semantics would be. But before we start designing API flags, we need to consider what races actually exist, what they do, and what we can do to mitigate them. That's a task for the MM group. But what I think I can fairly confidently say is that there isn't going to be a magic "no undefined behavior" flag.

**Comment by Robert-Qualcomm** [comment]
> What about declaring data races on external memory to not be undefined? Implementations could either assume that data races could happen everywhere, or only assume that data races will not happen if external memory is not involved.

Can we please use the correct terminology here. Data races are UB by definition. What you actually mean is to declare all external memory **conflicts** to be atomic. That is certainly a possibility. The only caveat is that in theory it might result in out-of-thin-air values but that is still the subject of research.

**Comment by DemiMarie** [comment]
> > What about declaring data races on external memory to not be undefined? Implementations could either assume that data races could happen everywhere, or only assume that data races will not happen if external memory is not involved.
> 
> Can we please use the correct terminology here. Data races are UB by definition. What you actually mean is to declare all external memory **conflicts** to be atomic. That is certainly a possibility. The only caveat is that in theory it might result in out-of-thin-air values but that is still the subject of research.

I’m not familiar with the “conflict” terminology.  Do you mean “access”?

> Yeah, on-the-fly re-compilation is right out. Device flags aren't great, either. Something at compile time might work but it's not clear what it would do and what the semantics would be. But before we start designing API flags, we need to consider what races actually exist, what they do, and what we can do to mitigate them. That's a task for the MM group. But what I think I can fairly confidently say is that there isn't going to be a magic "no undefined behavior" flag.

My goals are to provide a solution that:

1. Is easy for compositor authors to adopt.
2. Does not require annotating every access to client-provided buffers.
3. Provides guarantees instead of relying on a program with undefined behavior to happen to work.
4. Provides sufficiently _strong_ guarantees that defensive copies are almost needed in practice.  I expect that making every load and store byte-wise atomic should work, and I am not aware of a weaker guarantee that would be sufficient.  I am open to being convinced otherwise, though.

**Comment by mahkoh** [comment]
>Does not require annotating every access to client-provided buffers.

I don't want to suffer any performance overhead for this when dealing with safe external memory and I'm fine with adding new extension objects to every operation that involve dangerous memory and annotating shader operations that operate on dangerous memory.

**Comment by DemiMarie** [comment]
> > Does not require annotating every access to client-provided buffers.
> 
> I don't want to suffer any performance overhead for this when dealing with safe external memory and I'm fine with adding new extension objects to every operation that involve dangerous memory and annotating shader operations that operate on dangerous memory.

Do you have examples of safe external memory?  I can’t think of any, at least that involve ordinary clients (which I expect to be the vast majority).

My concern is that most compositors will not do this in practice, especially because this is _not_ a problem that testing can catch.  The validation layers could conservatively assume that _all_ external memory is unsafe, but otherwise (like other time-of-check to time-of-use bugs) this is something that could pass every possible test involving legitimate clients and then explode under deliberate attack.

That’s why I want something that happens almost entirely in the implementation: it is much more likely for one implementation to get it right than for every compositor out there to annotate all of their accesses correctly.

**Comment by mahkoh** [comment]
>Do you have examples of safe external memory?

Buffers that I allocate myself with GBM. gamescope probably doesn't care at all because it deals with a single client. Applications that import buffers from vaapi which already has full access to the application's memory.

**Comment by DemiMarie** [comment]
> > Do you have examples of safe external memory?
> 
> Buffers that I allocate myself with GBM. gamescope probably doesn't care at all because it deals with a single client. Applications that import buffers from vaapi which already has full access to the application's memory.

To me, it would be much better to treat external memory as unsafe by default, and require annotations to mark it as safe.  That follows the secure-by-default principle.

What I am trying to avoid is a situation where sandboxing platforms are forced to make defensive copies in order to protect the compositor.  While I have no data, my understanding is that the cost of this would dwarf that of any of the other options proposed here.

**Comment by gfxstrand** [comment]
I'm going to repeat myself here: External memory is not special. It's just memory that happens to be shared. There's nothing we can add to external memory as such. All these concepts of "safe memory" or "safe by default on external memory" are nonsense. It's all just memory. Anything we would add would be to individual accesses to constrain the potential damage.

Could that be a big switch to make it easier to turn on so it's easy for compositor developers? Possibly. It depends on the nature of the MM changes. (Which we still don't know what they are yet.) There may not even need to be a switch for some of it. But what I can 100% tell you is that you're not getting a different memory model on external memory.

**Comment by Tobski** [comment]
The memory model discussion feels to me like a massive red herring. There's nothing stopping _any_ UB in Vulkan from having the same effect as a data race would - what is being asked for here is both technically infeasible (making all external memory reads safe) and probably doesn't actually solve the problem (UB causing things to blow up).

If we want to make stronger guarantees about what happens when _any_ UB occurs, then we can do that in an extension. UB does not have to include spawning a black hole at your desk; we just don't spec that yet.

**Comment by Tobski** [comment]
Right now, FWIW, we have this _fairly_ wishy washy statement about UB in Vulkan:

> The core layer assumes applications are using the API correctly. Except as documented elsewhere in the Specification, the behavior of the core layer to an application using the API incorrectly is undefined, and may include program termination. However, implementations must ensure that incorrect usage by an application does not affect the integrity of the operating system, the Vulkan implementation, or other applications in the system using Vulkan.

Fundamentally though, this is not tested, and it was written before the external memory extensions got involved, but it's the core of something useful. If we wanted to verify this fully, it would probably mean stress testing almost every single valid usage statement to make sure the OS didn't explode after running a test, and I don't think that's likely to happen unless someone wants to spend several million $ on this.

However; we could probably at least _consider_ something a bit more modest in scope, limited to specific UB conditions involving shared memory.

**Comment by DemiMarie** [comment]
I think there is some confusion here, and honeslty much of it is my fault.  My focus is security and especially virtuailzation-based security, and the reason I filed this issue is because certain parts of Vulkan are responsible for _implementing_ the guarantees @Tobski noted above.  However, the phrasing I used was definitely poor.

The reason I filed this issue is that I am not aware of good solution for writing compositors in Vulkan.  In particular, wlroots (whose authors know far more about Vulkan than I ever will) _does not_ use atomic operations to access client-provided buffers.  This means that either they were unaware of the potential risks, or they decided that the alternatives were all worse.  I consider the latter to be far more likely.

From interacting with @valpackett and from reading https://github.com/KhronosGroup/Vulkan-Docs/issues/2408#issuecomment-3137730278, it appears that this is not something compositor authors are particularly worried about.  Therefore, I suspect solutions that require invasive refactorings (such as using atomic loads for all texture reads) are unlikely to be accepted.  Additionally, from a quick read of the Vulkan and SPIR-V specifications, it appeared that not all operations compositor authors might legitimately need to use have atomic versions.  Therefore, my interest is mostly in solutions that implicitly make _all_ loads and stores have bytewise-atomic semantics: they may tear, but the implementation cannot duplicate loads or assume that loads without an intervening store return the same value.  Enabling this with a device flag would allow for compositors to avoid time-of-check to time-of-use bugs in a non-invasive way, provided that the shaders do not have such bugs at the source code level.

**Comment by Tobski** [comment]
Ok so the problem here is writing a safe compositor - that was certainly not apparent before!

Firstly, Vulkan already has a huge array of tools to make accesses safe, if a compositor writer isn't choosing to use those tools, it's most likely because they have a fairly huge cost inherent to how they work - atomics are slower than regular accesses. There's not much we can do about that, and there's no magical device flag we could implement that would do anything except making everything incredibly slow.

That's the bad news.

The good news though is that we are working on better tools to deal with safety and eliminating UB; again none of these are a magical device switch, but we are working on making them easier to work with (and ideally faster!) than the tools we have today.

My conclusion then is that there's nothing really actionable we can do with this issue, unfortunately; it's asking for far too much. We will continue to work on and deliver continuous improvements to this area, but I don't know that we can ever get to the point where this is "solved". What would be much more useful is having issues raised by compositor authors on _why_ they aren't using the tools they have available (or any tools they have missing), because then maybe we can work on addressing those more specific issues.

We have a handful of items being worked on currently which will directly impact work in this area, so please do keep an eye on what we start shipping.

Also I would suggest taking a look at how LLVM handles data races, because they have much safer data races, and we're currently looking in that area to see what we can do in Vulkan.

**Comment by mahkoh** [comment]
>Firstly, Vulkan already has a huge array of tools to make accesses safe, if a compositor writer isn't choosing to use those tools, it's most likely because they have a fairly huge cost inherent to how they work - atomics are slower than regular accesses.

I don't believe there are such tools at the moment. Like I said, access has to happen through VkImage since the memory layout is in general unknown to the compositor  (and not knowable without hard coding the memory layouts for each device out there). And I believe the layout transitions required to use the images already count as non-atomic access.

**Comment by Tobski** [comment]
> > Firstly, Vulkan already has a huge array of tools to make accesses safe, if a compositor writer isn't choosing to use those tools, it's most likely because they have a fairly huge cost inherent to how they work - atomics are slower than regular accesses.
> 
> I don't believe there are such tools at the moment. Like I said, access has to happen through VkImage since the memory layout is in general unknown to the compositor (and not knowable without hard coding the memory layouts for each device out there). And I believe the layout transitions required to use the images already count as non-atomic access.

https://registry.khronos.org/vulkan/specs/latest/man/html/VK_KHR_unified_image_layouts.html seems relevant?

**Comment by mahkoh** [comment]
I had not considered that the compositor could always use the general layout with or without that extension. 

But I believe at least one transition from the undefined layout would be required and the compositor must perform a queue transfer from and to the foreign queue before and after each access.

**Comment by Robert-Qualcomm** [comment]
> > > What about declaring data races on external memory to not be undefined? Implementations could either assume that data races could happen everywhere, or only assume that data races will not happen if external memory is not involved.
> > 
> > 
> > Can we please use the correct terminology here. Data races are UB by definition. What you actually mean is to declare all external memory **conflicts** to be atomic. That is certainly a possibility. The only caveat is that in theory it might result in out-of-thin-air values but that is still the subject of research.
> 
> I’m not familiar with the “conflict” terminology. Do you mean “access”?

A conflict is a non-synchronized concurrent access involving at least one write. A key property of the memory model is that the user can specify whether a conflict should be allowed (by making it atomic),, or disallowed (a data race). If conflicts are disallowed, the compiler can assume no other thread will be accessing the memory location, which allows for greater optimization.

**Comment by Tobski** [comment]
> I had not considered that the compositor could always use the general layout with or without that extension.
> 
> But I believe at least one transition from the undefined layout would be required and the compositor must perform a queue transfer from and to the foreign queue before and after each access.

It's not clear to me why you as a compositor would transition to/from undefined after the app has access to the image.

Regardless, it seems like we're drilling down on something that might actually be useful to investigate - would you be happy to raise a separate issue on this?

**Comment by mahkoh** [comment]
>It's not clear to me why you as a compositor would transition to/from undefined after the app has access to the image.

I believe one transition of VkImage away from undefined is always required after creating the VkImage even if it is backed by imported memory.

>Regardless, it seems like we're drilling down on something that might actually be useful to investigate - would you be happy to raise a separate issue on this?

I believe that what I'm describing belongs in this issue since this issue is about avoiding undefined behavior (and more) in the presence of conflicting memory access. I don't know why other people focus so much on access in shaders when there the compositor at least has some amount of control wrt. how memory is accessed. I'm more worried about all of the other function calls that take a VkImage as a parameter.

**Comment by DemiMarie** [comment]
> Ok so the problem here is writing a safe compositor - that was certainly not apparent before!

I’m not surprised.  I’ve never even used Vulkan before 😆.  The reason I filed this issue is because I do plan to rely on code that _does_ use Vulkan (in the future if not already) and because reading existing compositors suggested that they were unsafe.

> Firstly, Vulkan already has a huge array of tools to make accesses safe, if a compositor writer isn't choosing to use those tools, it's most likely because they have a fairly huge cost inherent to how they work - atomics are slower than regular accesses. There's not much we can do about that, and there's no magical device flag we could implement that would do anything except making everything incredibly slow.

Would you be willing to go into more detail on this?

I am definitely opposed to actually generating hardware-level atomic instructions.  That _would_ be absurdly slow.  Rather, I would like compositors to be able to read from client-provided buffers in a way that safe, fast, _and_ easy to use.  The use-cases I can think of are also fine with eliminating redundant loads.  It’s _adding_ loads that could be problematic.

> That's the bad news.
> 
> The good news though is that we are working on better tools to deal with safety and eliminating UB; again none of these are a magical device switch, but we are working on making them easier to work with (and ideally faster!) than the tools we have today.
> 
> My conclusion then is that there's nothing really actionable we can do with this issue, unfortunately; it's asking for far too much.

I’m not surprised here either.  Feel free to close this issue in favor of more specific issues.

> We will continue to work on and deliver continuous improvements to this area, but I don't know that we can ever get to the point where this is "solved". What would be much more useful is having issues raised by compositor authors on _why_ they aren't using the tools they have available (or any tools they have missing), because then maybe we can work on addressing those more specific issues.

Not a compositor author myself (though pinging @emersion of wlroots and Sway), but I think at least one reason is that these problems are considered too theoretical to worth addressing (https://github.com/KhronosGroup/Vulkan-Docs/issues/2408#issuecomment-3137730278).  Even in security-critiical contexts, kernel driver vulnerabilities are (justifiably) considered a much bigger threat.

Another recommendation would be to provide examples of how to do this right.  From a quick look at the spec, it wasn’t at all obvious to me that atomic texture loads are even possible.  Also, no ordinary application has a need for this, so it won’t show up in tutorials.

> We have a handful of items being worked on currently which will directly impact work in this area, so please do keep an eye on what we start shipping.
> 
> Also I would suggest taking a look at how LLVM handles data races, because they have much safer data races, and we're currently looking in that area to see what we can do in Vulkan.

I think one difference is that languages like Rust and C++ allow encapsulating shared buffers in structs that don’t expose the underlying buffer directly, and instead provide accessors that use atomic operations internally.  This makes it _much_ easier for programmers to do the right thing, because if they try to do the wrong thing the compiler will complain.  I don’t know of any analog in the GPU world.

**Comment by Tobski** [comment]
> I believe that what I'm describing belongs in this issue since this issue is about avoiding undefined behavior (and more) in the presence of conflicting memory access. I don't know why other people focus so much on access in shaders when there the compositor at least has some amount of control wrt. how memory is accessed. I'm more worried about all of the other function calls that take a VkImage as a parameter.

I agree that discussing it here is perfectly fine, but my concern is that this issue is already rather long - and it seems like it's not stopping. I don't want this item to get lost in the noise.

**Comment by Tobski** [comment]
Maybe general point - issues like this are fine as long as we're having useful discussion. But if you want us (the Vulkan WG) to action anything, we need to have a well defined requests.

**Comment by DemiMarie** [comment]
> Maybe general point - issues like this are fine as long as we're having useful discussion. But if you want us (the Vulkan WG) to action anything, we need to have a well defined requests.

Here is one that is (hopefully) well defined:

For each load and store instruction, provide a version that is safe against concurrent access or modification, and which is expected to be almost as fast as the unsafe version.  This could mean something like:
- Implementations are allowed to coalese loads and stores.
- Implementations are _not_ allowed to insert spurious loads or stores.
- If there is concurrent modification the value read is unspecified.
- Implementations are expected to _not_ generate hardware-level atomic loads or stores, as those would be far too slow.  The only exception would be for hardware that actually requires them.  I (perhaps naively) expect no hardware to have this requirement.

High-level language compilers (GLSL, HLSL, others) could then provide a flag to generate these instructions instead of normal ones.

**Comment by Tobski** [comment]
> Here is one that is (hopefully) well defined:

Ah - to be clear, I was asking if these could be raised as separate issues - but since it's here I'll respond here for now.

> For each load and store instruction, provide a version that is safe against concurrent access or modification, and which is expected to be almost as fast as the unsafe version.

We're already looking into something like this, but it will not be "almost as fast" and there is fundamentally no way to make it "almost as fast" because almost all the things that make accesses fast are also what make them unsafe. Using these types of access in the general sense is going to be slow. Maybe not as slow as bypassing all the caches for a full system-level atomic would be, but still slow.

But to highlight what @mahkoh said above:

> I don't know why other people focus so much on access in shaders when there the compositor at least has some amount of control wrt. how memory is accessed. I'm more worried about all of the other function calls that take a VkImage as a parameter.

it doesn't sound like what you're asking for would move the needle anyway.

**Comment by DemiMarie** [comment]
> > For each load and store instruction, provide a version that is safe against concurrent access or modification, and which is expected to be almost as fast as the unsafe version.
> 
> We're already looking into something like this, but it will not be "almost as fast" and there is fundamentally no way to make it "almost as fast" because almost all the things that make accesses fast are also what make them unsafe. Using these types of access in the general sense is going to be slow. Maybe not as slow as bypassing all the caches for a full system-level atomic would be, but still slow.

Which things are these?  On the CPU such an access is just an ordinary non-atomic load instruction.  Are GPUs different?

**Comment by Tobski** [comment]
> Which things are these? On the CPU such an access is just an ordinary non-atomic load instruction. Are GPUs different?

The examples you gave as being "disallowed". The exact same is true on a CPU.

**Comment by DemiMarie** [comment]
> > Which things are these? On the CPU such an access is just an ordinary non-atomic load instruction. Are GPUs different?
> 
> The examples you gave as being "disallowed". The exact same is true on a CPU.

I thought that, in most cases, the code generated by shader compilers is _already_ safe against time-of-check to time-of-use, even if the specification doesn't require it.  In that case, there would be no performance penalty at all, because the compiler can generate the same code it did before.  The only case I can think of where adding extra loads (and thus creating potential time-of-check to time-of-use issues) is a win is where there is significant register pressure, and I thought compositor shaders generally didn't have that.

Is this reasoning incorrect?

**Comment by Tobski** [comment]
I'm not going to spend time unpicking your logic here - we're going to do a thing, if you want to track that then please raise your request as a separate issue. We have other use cases for this that don't rely on the specifics of how a compositor works, so once it exists compositor authors can figure out how/if they want to use it and then the impact can be measured.

**Comment by DemiMarie** [comment]
That makes sense.

**Comment by DemiMarie** [comment]
Closing for now.  I'll file a separate issue if I come up with anything specific.
